src/Agent.Plugins/BuildDrop/BuildDropPlugin.cs

```diff
@@ -0,0 +1,191 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.BlobStore.Common;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+using Microsoft.VisualStudio.Services.BlobStore.WebApi;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Agent.Sdk;
+
+namespace Agent.Plugins.BuildDrop
+{
+    public abstract class BuildDropTaskPluginBase : IAgentTaskPlugin
+    {
+        public abstract Guid Id { get; }
+        public string Version => "0.138.0"; // Publish and Download tasks will be always on the same version.
+        public string Stage => "main";
+
+        public async Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+
+            // Project ID
+            Guid projectId = new Guid(context.Variables.GetValueOrDefault(BuildVariables.TeamProjectId)?.Value ?? Guid.Empty.ToString());
```

Comment:
don't use BuildVariables.TeamProjectId, it might not exist in Release when you download drops.

Reply:
I modified this logic a little bit. For publish, I put check for build at the first so this should be available; for download, I am merely checking if "system.teamprojectid" is available, which happens to be the same value as BuildVariables.TeamProjectId, so I used the constant.